packages/react-reconciler/src/ReactFiberCompleteWork.new.js

```diff
@@ -1361,7 +1361,9 @@ function completeWork(
         const prevIsHidden = prevState !== null;
         if (
           prevIsHidden !== nextIsHidden &&
-          newProps.mode !== 'unstable-defer-without-hiding'
+          newProps.mode !== 'unstable-defer-without-hiding' &&
+          // LegacyHidden doesn't do any hiding — it only pre-renders.
```

Comment:
I'm confused. Legacy hidden _does_ toggle visibility, right? It's just not supposed to run layout effects when visibility changes.

Wouldn't removing this flag also prevent it from calling `hideOrUnhideAllChildren` later?

Reply:
It used to hide stuff but I removed that behavior, because some existing code in FB was relying on the ability to override the style of `<div hidden={true} />` with a CSS class by setting `display: block`. Which doesn't work if React sets an inline style.

So instead in www we always render LegacyHidden via a userspace abstraction that wraps it in a DOM node, and sets `hidden=true` on that DOM node. This also worked around the insertion problem fixed by this PR.

Reply:
The way the `hidden` attribute works is browser's have a built-in stylesheet with this rule:

```css
[hidden] {
  display: none;
}
```

That has pretty low precedence in CSS, so a user stylesheet can override it by setting `display` to something else.

But the way Offscreen/Suspense works is by setting the `display` property directly on the node with an inline style. We also use `!important`. So it can't be overridden.

We don't actually want to support overriding the style like this, but some code was already relying on it, so I added the LegacyHidden type to prevent a regression and unblock the Lanes refactor.

Reply:
> It used to hide stuff but I removed that behavior

I see. Looking at how the `Visibility` flag is used, it's what ensures `hideOrUnhideAllChildren` is called which in turn is what calls `hideInstance` or `unhideInstance` in the host config, so this change seemed like it would break that.

> So instead in www we always render LegacyHidden via a userspace abstraction that wraps it in a DOM node, and sets hidden=true on that DOM node. 

Ah, so basically FB is the only ones that will ever use this component, and we'll eventually clean it up there too? (So this API doesn't need to work outside of the FB context?)

Reply:
> Ah, so basically FB is the only ones that will ever use this component, and we'll eventually clean it up there too? (So this API doesn't need to work outside of the FB context?)

Yeah exactly

Reply:
> It used to hide stuff but I removed that behavior

To be clear I mean I removed this behavior like last year :D